airbyte-integrations/connectors/source-snapchat-marketing/source_snapchat_marketing/source.py

```diff
@@ -0,0 +1,397 @@
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return f"adaccounts/{stream_slice[self.slice_key_name]}/{self.response_root_name}"
+
+
+class Organizations(SnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#organizations """
+
+    def path(self, **kwargs) -> str:
+        return "me/organizations"
+
+
+class Adaccounts(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-ad-accounts """
+
+    depends_on_stream = Organizations
+    slice_key_name = "organization_id"
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return f"organizations/{stream_slice[self.slice_key_name]}/adaccounts"
+
+
+class Creatives(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-creatives """
+
+    depends_on_stream = Adaccounts
+
+
+class Media(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-media """
+
+    depends_on_stream = Adaccounts
+
+    @property
+    def response_item_name(self):
+        return self.response_root_name
+
+
+class Campaigns(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-campaigns """
+
+    depends_on_stream = Adaccounts
+
+
+class Ads(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-ads-under-an-ad-account """
+
+    depends_on_stream = Adaccounts
+
+
+class Adsquads(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-ad-squads-under-an-ad-account """
+
+    depends_on_stream = Adaccounts
+
+
+class Segments(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-audience-segments """
+
+    depends_on_stream = Adaccounts
+
+
+class SnapchatAdsOauth2Authenticator(Oauth2Authenticator):
+    """Request example for API token extraction:
+    curl -X POST https://accounts.snapchat.com/login/oauth2/access_token \
+      -d "refresh_token={refresh_token}" \
+      -d "client_id={client_id}" \
+      -d "client_secret={client_secret}"  \
+      -d "grant_type=refresh_token"  \
+    """
+
+    def __init__(self, config):
+        super().__init__(
+            token_refresh_endpoint="https://accounts.snapchat.com/login/oauth2/access_token",
+            client_id=config["client_id"],
+            client_secret=config["client_secret"],
+            refresh_token=config["refresh_token"],
+        )
+
+    def refresh_access_token(self) -> Tuple[str, int]:
+        response_json = None
+        try:
+            response = requests.request(method="POST", url=self.token_refresh_endpoint, data=self.get_refresh_request_body())
+            response_json = response.json()
+            response.raise_for_status()
+            return response_json["access_token"], response_json["expires_in"]
```

Comment:
We're not explicitly catching exceptions of this return so I think it should be in an `else:` making this a try/except/else block.

Reply:
Done. Move `return` block to `else`